ts/index.ts

```diff
@@ -1,17 +1,35 @@
 import StorageRegistry from './registry'
-import { StorageBackend, CreateSingleResult } from './types'
 import { createDefaultFieldTypeRegistry, FieldTypeRegistry } from './fields'
 export { default as StorageRegistry } from './registry'
+import {
+    StorageBackend,
+    CreateSingleOptions,
+    FindSingleOptions,
+    FindManyOptions,
+    SuggestOptions,
+    CountOptions,
+    UpdateManyOptions,
+    UpdateSingleOptions,
+    DeleteManyOptions,
+    DeleteSingleOptions,
+    CreateSingleResult,
+    SuggestResult,
+    DeleteSingleResult,
+    DeleteManyResult,
+    UpdateSingleResult,
+    UpdateManyResult,
+} from './types'
 
 export interface StorageCollection {
-    // putObject(object) : Promise<PutSingleResult>
-    createObject(object) : Promise<CreateSingleResult>
-    findOneObject<T>(query, options?) : Promise<T | null>
-    findObjects<T>(query, options?) : Promise<Array<T>>
-    updateOneObject(object, updates, options?)
-    updateObjects(query, updates, options?)
-    deleteOneObject(object, options?)
-    deleteObjects(query, options?)
+    createObject(object, options? : CreateSingleOptions) : Promise<CreateSingleResult>
+    findOneObject<T>(query, options?: FindSingleOptions) : Promise<T | null>
+    findObjects<T>(query, options?: FindManyOptions) : Promise<Array<T>>
+    countObjects(query, options?: CountOptions) : Promise<number>
+    suggestObjects<S, P = any>(query, options?: SuggestOptions) : Promise<SuggestResult<S, P>>
```

Comment:
Suggest seems like a search-specific operation, so I'd like to keep that out of both the standard Dexie back-end and the core of Storex. it's just a few standard queries that don't require any special ops to be done directly on Dexie, right? The search and tags related logic could be (an)other package(s).

Reply:
> [suggest is] just a few standard queries that don't require any special ops to be done directly on Dexie, right?

`suggestObjects` implementation uses Dexie's `WhereClause.startsWith` method to query on prefixes to a field rather than exact matches.

`findObjects` uses Dexie's `Collection.toArray` method to fetch all matching objects to return, while `suggestObjects` gets away with just interacting with the keys, which [is more performant than `Collection.toArray`](http://dexie.org/docs/Collection/Collection.primaryKeys()#performance-notice).

I would hope not to lose any performance from moving Memex to storex. 
Maybe we could create a `storex-extensions-memex` package and have a simple extensions system afforded in storex? Though I'm not sure how it should work architecturally. It would need to access the underlying storage backend, though that would only be available from the relevant `storex-backend-*` package. Maybe storex backend extensions? Any ideas?

Reply:
I'd just say that we have things like this in Memex's src/search dir, and we just pass in the underlying Dexie instance. We don't have to be religious about abstracting things  :)  Once we notice patterns and primitives that are useful for various use cases, we can standardize them.